### ISSUE TITLE
How to handle error from JSON.parse

### ISSUE THREAD

**Issue by mzabriskie** [axios/axios#61]
The `transformRequest` method attempts to parse JSON, but swallows any error in an empty `catch` statement. This results in the stringified form of the response body to be returned should the JSON be malformed, and supplied to the response object that is used to resolve the `Promise`.

Without the try/catch a global error would occur. With the try/catch the code handling the response will error, but in an unpredictable way. Consider:

``` js
// Assume a malformed response from the server for the request.
// Missing quotes around property names:
//   {firstName: "Jimmy", lastName: "Page", isBlocked: true}
axios.get('/user/12345').then((res) => {
  var user = res.data;
  if (!user.isBlocked) {
    // Do something potentially dangerous that only an unblocked user should be allowed to do
  }
});
```

In this case `user` is a `String` not an `Object` as expected. Calling `'some string'.isBlocked` will return `undefined`, which evaluates as falsey. This will cause the `if` condition to erroneously be entered.

A potential solution would be to reject the `Promise` if `JSON.parse` throws an `Error`. This isn't a straight forward fix though as it would break #55 again.

**Comment by jamesseanwright** [comment]
Personally, I think Promise rejection is a good solution. At the very least, one could invoke `console.error` in the `catch` block.

**Comment by demetriusnunes** [comment]
I also think there should be an easy option to retry the request in these cases as it often boils down to a truncated response.

**Comment by youurayy** [comment]
Just got burned on this. Instead of correctly rejecting the promise on invalid (truncated) JSON, it simply returns the orignal string object. Bad bad bad. At least put a warning in the docs about this.

**Comment by rubennorte** [comment]
I think we should just reject with the parse errors. That'd make axios more predictable and easier to debug.

**Comment by batiste** [comment]
I also spent 1 hour today tryng to understand what is happening with some invalid JSON. Just reject the Promise, or at least console.log something...?

**Comment by youurayy** [comment]
imagine first finding out in production - not sure where the logical thinking has gone on this one

**Comment by vinnymac** [comment]
Not sure how others are avoiding this issue after almost 3 years, but I decided to try to work around it for one of my projects. I would like to see how other people have been doing this as well.

I took notes from [here](https://github.com/sindresorhus/got/pull/301) and only raise an error for 2xx response codes.

If people find it useful I can make a PR of my [commit](https://github.com/vinnymac/axios/commit/067b75ed8ee04723af478ccc6c193adf911a5197#diff-5391bde944956870128be1136e7bc176). If you need this to raise for other response codes, just remove lines [67](https://github.com/vinnymac/axios/commit/067b75ed8ee04723af478ccc6c193adf911a5197#diff-5391bde944956870128be1136e7bc176R67) and [71](https://github.com/vinnymac/axios/commit/067b75ed8ee04723af478ccc6c193adf911a5197#diff-5391bde944956870128be1136e7bc176R71) in `defaults.js`

**Comment by thame** [comment]
I also encountered a scenario where this behavior is problematic. If the network connection is interrupted partially through the request, axios would not call an error and my application received malformed JSON which obviously led to lots of errors.

I too think that JSON parse errors should be caught and reported. I wasn't keen on forking or modifying the source files so my solution was to transform the response like this: 

```
axios.get(URL, {
  transformResponse: [(response) => (JSON.parse(response))], 
})
```

**Comment by jmkenz** [comment]
My approach to handling this is to wrap all axios requests in a promise, and reject that promise if the response from axios show a response.headers['content-type'] of 'application/json' and a response.data that is a string.  From this I infer there was a JSON parsing error.

```
return new Promise((resolve, reject) => {
  axios.request(options)
    .then(response => {
      if (response.headers && response.headers['content-type'] && response.headers['content-type'].indexOf('application/json') >= 0 && _.isString(response.data)) {
        reject(new Error('Error parsing response data as JSON object'));
      } else {
        resolve(response);
      }
  });
});
```

**Comment by youurayy** [comment]
@emilyemorehouse is this something you could fix, for the sake of all humanity?

**Comment by wujunze** [comment]
Should throw exception

**Comment by ethanentW** [comment]
This is among the most disappointing responses to a glaring issue I've seen in an open-source project. This has been an issue for almost 4 years... This is the reason I use my own HTTP package.

**Comment by todorone** [comment]
@vinnymac Can You please submit PR of Your solution?

**Comment by smakinson** [comment]
Just came across this as well, it seems if I manually add: `responseType: 'json'` (even though it's the default according to the docs) it at least keeps it form providing the bad data to the result.

**Comment by rognstad** [comment]
I just got bitten by this in production today. The current behavior is really unintuitive and wasted a bunch of my time in debugging. Throwing an error would have pointed me in the right direction immediately and would have been helpfully caught/grouped in our error reporting.

**Comment by yarcl** [comment]
Is this responseText is a problem?  like "24\r\n{\"success\":\"true\", \"payload\":\"123456\"}\r\n0\r\n\r\n", and sometimes it returns "24\r\n{\"success\":\"true\", \"payload\":\"123456\"}\r\n".

**Comment by meyve** [comment]
> imagine first finding out in production - not sure where the logical thinking has gone on this one

Generally speaking, that was exactly what we faced....

**Comment by porlov** [comment]
Guys, any updates on this issue?
This is really confusing behavior that I've faced with recently.

**Comment by jasonsaayman** [comment]
Fixed in #3688

**Comment by pcone** [comment]
I see this has been closed already, but imo neither of the available behaviours from silentJSONParsing are *quite* perfect.
With silentJSONParsing on you can access the status code of the response, but the JSON parsing error is swallowed which isn't ideal.
But with silentJSONParsing on there's no longer any way to see what the status code of the response was, unless I'm missing something.
